### PR TITLE
Provide a single service for adding imports.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This way we don't duplicate the code all over the place (for example, in 'add-using' and in 'snippet insertion').

Also, update Add-Using so that it attempts to add usings directives into groups of directives of the same type.  i.e. if you have code like:

```c#
using System;

namespace N
{
    // here are where my aliases go
    using Whatever = ...;
    
    class C
    {
         // Invoke 'Add Using' here
    }
}
```

In this case, we'll add the new using directives to the using-directive section in the compilation-unit, not to the alias-section in the inner namespace section.  This allows people to have different groups of using directives, for different purposes, without having add-using mess it up for them.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @dpoeschl

File: src/Workspaces/CSharp/Portable/AddImport/CSharpAddImportService.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.AddImport
+{
+    [ExportLanguageService(typeof(IAddImportService), LanguageNames.CSharp), Shared]
+    internal class CSharpAddImportService : IAddImportService
```
Comment (dpoeschl) [inline src/Workspaces/CSharp/Portable/AddImport/CSharpAddImportService.cs]: Do all the use cases of this service happen in C# land? I'd feel better if the VB one existed, even if it's less useful.

File: src/EditorFeatures/CSharpTest/Diagnostics/AddUsing/AddUsingTests.cs [hunk 2]
```diff
@@ -4373,6 +4373,43 @@ class C
 compareTokens: false);
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddUsing)]
+        public async Task TestPlaceUsingWithUsings_NotWithAliases()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/AddUsing/AddUsingTests.cs]: Yes and yes i believe.

File: src/Features/CSharp/Portable/CodeFixes/AddImport/CSharpAddImportCodeFixProvider.cs [hunk 3]
```diff
@@ -581,15 +553,19 @@ private NameSyntax CreateNameSyntax(IReadOnlyList<string> namespaceParts, int in
...
             SyntaxNode contextNode)
         {
-            var namespaceToAddTo = GetFirstContainingNamespaceWithUsings(contextNode);
-            var usingDirectives = namespaceToAddTo?.Usings ?? root.Usings;
+            var addImportService = document.GetLanguageService<IAddImportService>();
 
             var nameSyntax = namespaceOrTypeSymbol.GenerateNameSyntax();
+            var dummyUsing = SyntaxFactory.UsingDirective(nameSyntax);
+
+            var container = addImportService.GetImportContainer(root, contextNode, dummyUsing);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/AddImport/CSharpAddImportCodeFixProvider.cs]: Definitely.

File: src/Workspaces/CSharp/Portable/AddImport/CSharpAddImportService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/AddImport/CSharpAddImportService.cs]: Will do.

File: src/Workspaces/CSharp/Portable/AddImport/CSharpAddImportService.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        private static readonly Func<UsingDirectiveSyntax, bool> s_isUsing = u => u.Alias == null;
+        private static readonly Func<UsingDirectiveSyntax, bool> s_isAlias = u => u.Alias != null;
+
+        private static readonly Func<SyntaxNode, bool> s_hasAliases = n => GetUsings(n).Any(s_isAlias);
+        private static readonly Func<SyntaxNode, bool> s_hasUsings = n => GetUsings(n).Any(s_isUsing);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/AddImport/CSharpAddImportService.cs]: ehhhhhhhhhhh

File: src/Workspaces/Core/Portable/AddImport/IAddImportService.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+{
+    internal interface IAddImportService : ILanguageService
+    {
+        SyntaxNode AddImports(
+            SyntaxNode root, SyntaxNode contextLocation, 
+            IEnumerable<SyntaxNode> newImports, bool placeSystemNamespaceFirst);
+
+        /// <summary>
+        /// Given a context location in a provided syntax tree, returns the appropriate container
+        /// that <paramref name="import"/> should should be added to.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/AddImport/IAddImportService.cs]: should should should should what?

**Pull request merged**